pkg/squashfs/verity.go

```diff
@@ -200,7 +202,116 @@ func verityName(p string) string {
 	return fmt.Sprintf("%s-%s", p, veritySuffix)
 }
 
-func Mount(squashfs string, mountpoint string, rootHash string) error {
+func fileChanged(a os.FileInfo, path string) bool {
+	b, err := os.Lstat(path)
+	if err != nil {
+		return true
+	}
+	return !os.SameFile(a, b)
+}
+
+// Mount a filesystem as container root, without host root
+// privileges.  We do this using squashfuse.
+func GuestMount(squashFile string, mountpoint string) error {
+	if isMountpoint(mountpoint) {
+		return fmt.Errorf("%s is already mounted", mountpoint)
+	}
+
+	abs, err := filepath.Abs(squashFile)
+	if err != nil {
+		return errors.Errorf("Failed to get absolute path for %s: %v", squashFile, err)
+	}
+	squashFile = abs
+
+	abs, err = filepath.Abs(mountpoint)
+	if err != nil {
+		return errors.Errorf("Failed to get absolute path for %s: %v", mountpoint, err)
+	}
+	mountpoint = abs
+
+	cmdPath := findSquashfusePath()
+	if cmdPath == "" {
+		return fmt.Errorf("no squashfs binary found")
+	}
+	fiPre, err := os.Lstat(mountpoint)
+	if err != nil {
+		return fmt.Errorf("Failed stat'ing %q: %w", mountpoint, err)
+	}
+	if fiPre.Mode()&os.ModeSymlink != 0 {
+		return fmt.Errorf("Refusing to mount onto a symbolic linkd")
+	}
+
+	cmd, err := squashFuse("", squashFile, mountpoint)
+	if err != nil {
+		return err
+	}
+	err = cmd.Process.Release()
+	if err != nil {
+		return fmt.Errorf("Failed releasing squashfuse process: %w", err)
+	}
+
+	for count := 0; !fileChanged(fiPre, mountpoint); count++ {
+		if count%10 == 0 {
+			log.Debugf("%s is not yet mounted...\n", mountpoint)
+		}
+		time.Sleep(time.Duration(100 * time.Millisecond))
+	}
+
+	return nil
+}
+
+func isMountpoint(dest string) bool {
+	mounted, err := mount.IsMountpoint(dest)
+	return err == nil && mounted
+}
+
+// Takes /proc/self/uid_map contents as one string
+// Returns true if this is a uidmap representing the whole host
+// uid range.
```

Comment:
oh! I thought you were checking to see if `mappedUidZero == HostUidZero` . 

I assumed that operating as HostUidZero was enough.

Here is a script I wrote that shows it is enough, at least for ext2 and if root owns the block device (or the group owner is also mapped in).

```bash
#!/bin/sh

MP=""
LODEV=""
TEMP_D=""
RESTORE_LODEV=""
INSIDE="false"

fail() { msg "$@"; exit 1; }
msg() { echo "$@" 1>&2; }
cleanup() {
    msg "cleaning up"
    if [ -n "$MP" ]; then
        r umount "$MP"
    fi
    if [ -n "$LODEV" ]; then
        r losetup -d "$LODEV"
    fi
    if [ -n "$RESTORE_LODEV" ]; then
        r chown "$RESTORE_LODEV" "$LODEV"
    fi
    if [ -n "$TEMP_D" ]; then
        r rm -Rf "$TEMP_D"
    fi
}

r() {
    local p="$"
    [ "$INSIDE" = "true" ] && p="%"
    msg "$p" "$@"
    "$@"
}

inside() {
    local dev="$1" mp="$2"
    INSIDE=true
    msg "Inside: dev=\"$dev\" mp=\"$mp\""
    r id || fail
    r ls -l "$dev" &&
        r ls -ld "$mp" || fail "strange, ls failed"
    r mount "$dev" "$mp" || fail "mount failed"
    MP="$mp"
    r touch "$mp/file1"
    r ls "$mp/file1"
}

trap cleanup EXIT
if [ "$1" = "inside" ]; then
    shift
    inside "$@"
    exit
fi
TEMP_D=$(mktemp -d)

[ "$(id -u)" = "0" ] || fail "gotta be root"
echo "Outside:"
id || fail

r grep root /etc/subuid || fail
r grep root /etc/subgid || fail

img="${TEMP_D}/fs.img"
mp="${TEMP_D}/mp"
mkdir "$mp" || fail "failed mkdir mp"
r truncate --size=100M "$img" || fail "failed create image $img"

r mkfs.ext2 "$img" >/dev/null 2>&1 || fail "failed mkfs $img"

LODEV=$(r losetup --find --show "$img") || fail "failed losetup"
msg "created lodev $LODEV"

olduid=$(stat "--format=%U" "$LODEV") || fail "faild to get owner of $LODEV"
oldgid=$(stat "--format=%G" "$LODEV") || fail "faild to get owner group of $LODEV"
chown 0:0 "$LODEV" || fail "failed chown 0:0 $LODEV"
RESTORE_LODEV="$olduid:$oldgid"

r "$0" inside "$LODEV" "$mp" || fail "didn't work"
```

output looks like this:

```
$ sudo /tmp/try-me 
Outside:
uid=0(root) gid=0(root) groups=0(root),130(sbuild)
$ grep root /etc/subuid
root:0:1
$ grep root /etc/subgid
root:0:1
$ truncate --size=100M /tmp/tmp.6BqyRL5rEI/fs.img
$ losetup --find --show /tmp/tmp.6BqyRL5rEI/fs.img
created lodev /dev/loop14
$ /tmp/try-me inside /dev/loop14 /tmp/tmp.6BqyRL5rEI/mp
Inside: dev="/dev/loop14" mp="/tmp/tmp.6BqyRL5rEI/mp"
% id
uid=0(root) gid=0(root) groups=0(root),130(sbuild)
% ls -l /dev/loop14
brw-rw---- 1 root root 7, 14 Dec 19 22:24 /dev/loop14
% ls -ld /tmp/tmp.6BqyRL5rEI/mp
drwxr-xr-x 2 root root 4096 Dec 19 22:24 /tmp/tmp.6BqyRL5rEI/mp
% mount /dev/loop14 /tmp/tmp.6BqyRL5rEI/mp
% touch /tmp/tmp.6BqyRL5rEI/mp/file1
% ls /tmp/tmp.6BqyRL5rEI/mp/file1
/tmp/tmp.6BqyRL5rEI/mp/file1
cleaning up
% umount /tmp/tmp.6BqyRL5rEI/mp
cleaning up
$ losetup -d /dev/loop14
$ chown root:disk /dev/loop14
$ rm -Rf /tmp/tmp.6BqyRL5rEI
```

Reply:
I'm missing something - where are you doing a userns unshare and mapping?  I see no lxc-usernsexec or unshare call in there.

If I edit your script to do:
```
-r "$0" inside "$LODEV" "$mp" || fail "didn't work"
+r unshare -U -r "$0" inside "$LODEV" "$mp" || fail "didn't work"
```
Then it fails with
```
% mount /dev/loop32 /tmp/tmp.oZx0FfP99s/mp
mount: /tmp/tmp.oZx0FfP99s/mp: permission denied.
mount failed
```
Now, it *is* interesting that losetup succeeded.  I didn't expect that.

Reply:
bah, you're right. the last line of the script was supposed to say:
```r lxc-usernsexec -m b:0:0:1 -- "$0" inside "$LODEV" "$mp" || fail "didn't work"```

so yeah, that doesn't work. But why not?

Reply:
Because to have capable(CAP_SYS_ADMIN), which equals ns_capable(init_user_ns, CAP_SYS_ADMIN), you (a) must have CAP_SYS_ADMIN in your pE (which you do), and (b) must be in a user_ns which is same as or ancestor to the target user_ns, init_user_ns.  Your userns is a child of it.  You cannot obtain capable(CAP_ANYTHING).